### PR TITLE
[convection diffusion application] Add evaluation of the RHS in CalculateLocalSystem with external force /= 0

### PR DESCRIPTION
**Pull request by riccardotosi**
Add the evaluation of the righthandside in LaplacianElement::CalculateLocalSystem.

### CODE REVIEW

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 1]
```diff
@@ -92,12 +101,16 @@ void LaplacianElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, Vec
...
+        
         double IntToReferenceWeight = integration_points[PointNumber].Weight() * DetJ0;
         noalias(rLeftHandSideMatrix) += IntToReferenceWeight * prod(DN_DX, trans(DN_DX)); //
+
+        // Calculating the local RHS
+        noalias(rRightHandSideVector) += heat_flux_local * IntToReferenceWeight;
     }
-    //calculating external forces
-    noalias(rRightHandSideVector) = ZeroVector(number_of_points); //case of zero ext forces
+    // //calculating external forces
```
Comment (loumalouomega) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]:
```suggestion
```

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 2]
```diff
@@ -92,12 +101,16 @@ void LaplacianElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, Vec
...
+    // noalias(rRightHandSideVector) = ZeroVector(number_of_points); //case of zero ext forces
```
Comment (loumalouomega) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]:
```suggestion
```

Avoid commented code please

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 3]
```diff
@@ -92,12 +101,16 @@ void LaplacianElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, Vec
 
         //Calculating the cartesian derivatives (it is avoided storing them to minimize storage)
         noalias(DN_DX) = prod(DN_De[PointNumber],InvJ0);
-
+        
         double IntToReferenceWeight = integration_points[PointNumber].Weight() * DetJ0;
```
Comment (loumalouomega) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]:
```suggestion
        const double IntToReferenceWeight = integration_points[PointNumber].Weight() * DetJ0;
```

**Review by loumalouomega — COMMENTED** [review]
I added some comments

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 4]
```diff
@@ -13,7 +13,7 @@
 
 
 // External includes
-
+#include "utilities/geometry_utilities.h"
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: this element should work with arbitrary geometries, whily geometry_utilities is only for triangle and tets...please do not include this here

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 5]
```diff
@@ -76,14 +76,23 @@ void LaplacianElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, Vec
...
+
     Element::GeometryType::JacobiansType J0;
     Matrix DN_DX(number_of_points,dim);
     Matrix InvJ0(dim,dim);
     Vector temp(number_of_points);
+    
+    Vector heat_flux_local(number_of_points);
+    for(unsigned int node_element = 0; node_element<number_of_points; node_element++)
+    {
+        heat_flux_local[node_element] = GetGeometry()[node_element].GetSolutionStepValue(HEAT_FLUX) * N_local(0,node_element);
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: use "FastGetSolutionStepValue" 

also when you gather this value do not multiply it by N (and specially NOT by the N which corresponds to the first gauss point)

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 6]
```diff
@@ -92,12 +101,14 @@ void LaplacianElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, Vec
...
         //Calculating the cartesian derivatives (it is avoided storing them to minimize storage)
         noalias(DN_DX) = prod(DN_De[PointNumber],InvJ0);
-
-        double IntToReferenceWeight = integration_points[PointNumber].Weight() * DetJ0;
+        
+        const double IntToReferenceWeight = integration_points[PointNumber].Weight() * DetJ0;
         noalias(rLeftHandSideMatrix) += IntToReferenceWeight * prod(DN_DX, trans(DN_DX)); //
+
+        // Calculating the local RHS
+        noalias(rRightHandSideVector) += heat_flux_local * IntToReferenceWeight;
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: this is not correct. you should do

        auto N = row(Nlocal,PointNumber); //these are the N which correspond to the gauss point "PointNumber" (bad variable name btw)
        qgauss = inner_prod(N, heat_flux_local);
        noalias(rRighthandSideVector) += qgauss*N;

also note that you must initialize
 
     rRightHandSide to zero outside of the loop

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 7]
```diff
@@ -76,28 +75,41 @@ void LaplacianElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, Vec
     //reading integration points and local gradients
     const GeometryType::IntegrationPointsArrayType& integration_points = GetGeometry().IntegrationPoints();
     const GeometryType::ShapeFunctionsGradientsType& DN_De = GetGeometry().ShapeFunctionsLocalGradients();
-    
+    const Matrix& N_gausspoint = GetGeometry().ShapeFunctionsValues();
```
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: mmm
this is the Ns on all the gauss points. use the plural in the name or a better name. asideo fo this looks ok to me, thx

**Review by RiccardoRossi — DISMISSED** [review]

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp (hunk 7 above)
Comment (riccardotosi) [reply applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: Name changed.

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 8]
```diff
@@ -76,28 +75,41 @@ void LaplacianElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, Vec
...
-
-        double IntToReferenceWeight = integration_points[PointNumber].Weight() * DetJ0;
+        
+        const double IntToReferenceWeight = integration_points[PointNumber].Weight() * DetJ0;
         noalias(rLeftHandSideMatrix) += IntToReferenceWeight * prod(DN_DX, trans(DN_DX)); //
+
+        // Calculating the local RHS
+        double qgauss;
+        auto N = row(N_gausspoints,PointNumber); //these are the N which correspond to the gauss point "PointNumber"
+        qgauss = inner_prod(N, heat_flux_local);
```
Comment (philbucher) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: const double qgauss = ...

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp [hunk 9]
```diff
@@ -76,28 +75,41 @@ void LaplacianElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, Vec
...
-    
+    const Matrix& N_gausspoints = GetGeometry().ShapeFunctionsValues();
+
     Element::GeometryType::JacobiansType J0;
     Matrix DN_DX(number_of_points,dim);
     Matrix InvJ0(dim,dim);
     Vector temp(number_of_points);
 
+    Vector heat_flux_local(number_of_points);
+    for(unsigned int node_element = 0; node_element<number_of_points; node_element++)
```
Comment (philbucher) [inline applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: for (std::size_t i_node=0; i_node<numpoints; ++i_node) {
...
}

**Review by philbucher — COMMENTED** [review]
Some comments

File: applications/convection_diffusion_application/custom_elements/laplacian_element.cpp (hunk 9 above)
Comment (riccardotosi) [reply applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: Thanks for the two hints!

Comment (RiccardoRossi) [reply applications/convection_diffusion_application/custom_elements/laplacian_element.cpp]: you did not correct either...

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**